### ISSUE TITLE
Fix S2583/S2589 FP: Rule S4158 sets number constraint on Length property, unrelated to Collections.

### ISSUE THREAD

**Issue by omarjuul** [SonarSource/sonar-dotnet#8045]
### Description

See repro. I do not get what is going on here.

### Repro steps

``` csharp
class TestFP
{
    private int Length { get; }

    internal bool Test()
    {
        Dictionary<string, int> foo = new()
        {
            { "foz", 1 },
            { "baz", 42 },
        };
        bool result = foo.ContainsKey("foo");

        if (this.Length < 0) // S2589 triggers here
        {
            result = false;
        }

        return result;
    }
}
```

### Expected behavior

No trigger on rule S2589 expected.

### Actual behavior

Rule S2589 triggers, for some reason.

### Known workarounds

We could suppress the warning, or stay on version 9.7 of SonarAnalyzer (which is what we will do for now).

### Related information

* SonarAnalyzer version `9.10.0.77988` (False Positive not present in version `9.7`)

**Comment by mary-georgiou-sonarsource** [comment]
Hello @omarjuul,
Thank you for reporting this.
I confirm this is an FP and we should fix it in a next hardening.
